src/basic/terminal-util.c

```diff
@@ -230,31 +231,245 @@ int ask_char(char *ret, const char *replies, const char *fmt, ...) {
+         * STDIN_FILENO and STDOUT_FILENO here, but let's be overly correct for once, after all libc allows
+         * swapping out stdin/stdout. */
+        int fd_input = fileno(stdin);
+        int fd_output = fileno(stdout);
+        if (fd_input < 0 || fd_output < 0 || same_fd(fd_input, fd_output) <= 0)
+                goto fallback;
+
+        /* Try to disable echo, which also tells us if this even is a terminal */
+        struct termios old_termios;
+        if (tcgetattr(fd_input, &old_termios) < 0)
+                goto fallback;
+
+        struct termios new_termios = old_termios;
+        termios_disable_echo(&new_termios);
+        if (tcsetattr(fd_input, TCSADRAIN, &new_termios) < 0)
+                return -errno;
+
+        for (;;) {
+                int c = fgetc(stdin);
+
+                /* On EOF or NUL, end the request, don't output anything anymore */
+                if (IN_SET(c, EOF, 0))
+                        break;
+
+                /* On Return also end the request, but make this visible */
+                if (IN_SET(c, '\n', '\r')) {
+                        fputc('\n', stdout);
+                        break;
+                }
+
+                if (c == '\t') {
+                        /* Tab */
+
+                        _cleanup_strv_free_ char **completions = NULL;
+                        if (get_completions) {
+                                r = get_completions(string, &completions, userdata);
+                                if (r < 0)
+                                        goto fail;
+                        }
+
+                        _cleanup_free_ char *new_string = NULL;
+                        CompletionResult cr = pick_completion(string, completions, &new_string);
+                        if (cr < 0) {
+                                r = cr;
+                                goto fail;
+                        }
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_FULL)) {
+                                /* Output the new suffix we learned */
+                                fputs(ASSERT_PTR(startswith(new_string, strempty(string))), stdout);
+
+                                /* And update the whole string */
+                                free_and_replace(string, new_string);
+                                n = strlen(string);
+                        }
+                        if (cr == COMPLETION_NONE)
+                                fputc('\a', stdout); /* BEL */
+
+                        if (IN_SET(cr, COMPLETION_PARTIAL, COMPLETION_ALREADY)) {
+                                /* If this worked only partially, or if the user hit TAB even though we were
+                                 * complete already, then show the full list. */
+                                fputc('\n', stdout);
+
+                                _cleanup_free_ char **filtered = strv_filter_prefix(completions, string);
```

Comment:
`_cleanup_free_` -> `_cleanup_strv_free_`

When COMPLETION_ALREADY, this should provide only one element. Is it expected? Or do I misunderstand something?

Reply:
yes, that's expected. It's how we visualize that there's only one completion available. I think this behaves quite nicely.

I had this differently earlier, hence the bogus comment. But it felt wrong. I forgot to update the comment. will fix.